lib/fizzy/capi.cpp

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "execute.hpp"
+#include "instantiate.hpp"
+#include "parser.hpp"
+#include <fizzy/fizzy.h>
+#include <memory>
+
+namespace
+{
+/// Simple implementation of C++20's std::bit_cast.
+template <typename DstT, typename SrcT>
+DstT bit_cast(SrcT x) noexcept
+{
+    DstT z;
+    static_assert(sizeof(x) == sizeof(z));
+    __builtin_memcpy(&z, &x, sizeof(x));
+    return z;
+}
+}  // namespace
+
+extern "C" {
+bool fizzy_validate(const uint8_t* wasm_binary, size_t wasm_binary_size)
+{
+    try
+    {
+        fizzy::parse({wasm_binary, wasm_binary_size});
+        return true;
+    }
+    catch (...)
+    {
+        return false;
+    }
+}
+
+struct fizzy_module
+{
+    fizzy::Module module;
+};
+
+fizzy_module* fizzy_parse(const uint8_t* wasm_binary, size_t wasm_binary_size)
+{
+    try
+    {
+        auto cmodule = std::make_unique<fizzy_module>();
+        cmodule->module = fizzy::parse({wasm_binary, wasm_binary_size});
+        return cmodule.release();
+    }
+    catch (...)
+    {
+        return nullptr;
+    }
+}
+
+void fizzy_free_module(fizzy_module* module)
+{
+    delete module;
+}
+
+struct fizzy_instance
+{
+    fizzy::Instance* instance;
+};
+
+fizzy_instance* fizzy_instantiate(fizzy_module* module,
+    const fizzy_external_function* imported_functions, uint32_t imported_functions_size)
+{
+    try
+    {
+        std::vector<fizzy::ExternalFunction> functions(imported_functions_size);
+        for (size_t imported_func_idx = 0; imported_func_idx < imported_functions_size;
+             ++imported_func_idx)
+        {
+            auto func = [cfunc = imported_functions[imported_func_idx]](fizzy::Instance& instance,
+                            fizzy::span<const fizzy::Value> args,
+                            int depth) -> fizzy::ExecutionResult {
+                fizzy_instance cinstance{&instance};
+                const auto cres = cfunc.function(cfunc.context, &cinstance,
+                    reinterpret_cast<const fizzy_value*>(args.data()),
+                    static_cast<uint32_t>(args.size()), depth);
+
+                if (cres.trapped)
+                    return fizzy::Trap;
+                else if (!cres.has_value)
+                    return fizzy::Void;
+                else
+                    return bit_cast<fizzy::Value>(cres.value);
+            };
+
+            // TODO get type from input array
+            auto func_type = module->module.imported_function_types[imported_func_idx];
+
+            functions[imported_func_idx] =
+                fizzy::ExternalFunction{std::move(func), std::move(func_type)};
+        }
+
+        auto instance = fizzy::instantiate(std::move(module->module), std::move(functions));
+
+        auto cinstance = std::make_unique<fizzy_instance>();
+        cinstance->instance = instance.release();
+
+        fizzy_free_module(module);
+        return cinstance.release();
+    }
+    catch (...)
+    {
+        return nullptr;
```

Comment:
This is the annoying part, if this function consumes `module`, I think it should consume it in every case. This means `free_module` should be called here.

Reply:
Yeah, agree. It's a bit weird that function modifies it's argument in case of failure, but it reflects the way it is currently in C++ API: `fizzy::instantiate(std::move(module))` would leave `module` empty even if it fails.

On the other hand, in C++ the caller has the freedom to copy the module before instantiate if they want to. Maybe we should add `fizzy_copy_module` to make this an option in C API, too.

Reply:
Yeah I thought about duplication too.

Reply:
> This is the annoying part, if this function consumes `module`, I think it should consume it in every case. This means `free_module` should be called here.

Another option could be: `instantiate` invalidates `module`, but user is still required to call `fizzy_free_module` in the end.